lib/multipipe/multi_pipe.go

```diff
@@ -0,0 +1,120 @@
+package multipipe
+
+import (
+	"io"
+	"sync"
+)
+
+// TODO: reduce lock contention if it becomes a bottleneck
+
+// MultiPipe is an io.Writer that can create multiple readers from its contents
+type MultiPipe struct {
+	buf    []byte
+	rdErr  error
+	closed bool
+	cond   *sync.Cond
+}
+
+// Reader is an io.Reader that reads from the beginning of its parent MultiPipe
+// without affecting the other readers
+type Reader struct {
+	parent *MultiPipe
+	offset int
+}
+
+// CopyPipe copies all the bytes from r to w without blocking, returns a channel
+// that will be closed on completion
+func CopyPipe(r io.Reader, w *MultiPipe) chan struct{} {
+	ch := make(chan struct{})
+
+	go func() {
+		_, err := io.Copy(w, r)
+		w.CloseWithError(err)
+		close(ch)
+	}()
+
+	return ch
+}
+
+// Read reads all the available contents from the MultiPipe parent, then if
+// there is a read error or the stream is closed, an error will be returned
+func (m *Reader) Read(p []byte) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	m.parent.cond.L.Lock()
+	defer m.parent.cond.L.Unlock()
+
+	if !m.parent.closed || m.offset >= len(m.parent.buf) {
+		m.parent.cond.Wait()
+	}
```

Comment:
What if the parent MultiReader is already closed but the current offset is less than the length of the buffer? Will this block forever without allowing you to read the rest of the contents?

Reply:
If I understood you correctly you are stating that:

1. _parent MultiReader is already closed_, so that means: `m.parent.closed == true`
1. _current offset is less than the length of the buffer_, meaning that: `m.offset >= len(m.parent.buf) == false`

In that case, the condition `!m.parent.closed || m.offset >= len(m.parent.buf)` evaluates to `!true || false` == `false || false` == `false`, so the code won't call `m.parent.cond.Wait()` and hence won't block.

Reply:
Okay, I see what you are saying now, I'll fix that condition.